bin/reth-benchmark/src/block_fetcher.rs

```diff
@@ -0,0 +1,254 @@
+use tracing::error;
+
+/// This implements a stream of blocks for the benchmark mode
+pub(crate) struct BlockFetcher<'a, T> {
+    /// The mode of the benchmark
+    mode: BenchmarkMode,
+    /// The provider to be used to fetch blocks
+    provider: &'a RootProvider<T>,
+    /// The sending channel for blocks
+    block_sender: PollSender<TransportResult<Option<Block>>>,
+    /// The current block future in flight
+    current_block: BoxFuture<'a, TransportResult<Option<Block>>>,
+    // TODO: make this include a fn, with generic output, instead of just
+    // TransportResult<Option<Block>>.
+    //
+    // Then the fn can be used to create not only a stream of blocks, but return for example the
+    // corresponding FCU hashes as well.
+    //
+    // i.e. the output can be (Block, head, safe, finalized) or something like that.
+    //
+    // And this could be configured by the user
+}
+
+impl<'a, T> BlockFetcher<'a, T>
+where
+    T: Transport + Clone,
+{
+    /// Create a new `BlockStream` with the given mode and provider.
+    #[allow(clippy::type_complexity)]
+    pub(crate) fn new(
+        mut mode: BenchmarkMode,
+        provider: &'a RootProvider<T>,
+        buffer_size: usize,
+    ) -> Result<(Self, Receiver<TransportResult<Option<Block>>>), BlockFetcherError> {
+        let (block_sender, block_receiver) = channel(buffer_size);
+        let current_block = match mode {
+            BenchmarkMode::Continuous => {
+                // fetch Latest block
+                provider.get_block_by_number(BlockNumberOrTag::Latest, true)
+            }
+            BenchmarkMode::Range(ref mut range) => {
+                match range.next() {
+                    Some(block_number) => {
+                        // fetch first block in range
+                        provider.get_block_by_number(block_number.into(), true)
+                    }
+                    None => {
+                        // return an error
+                        return Err(BlockFetcherError::RangeEmpty);
+                    }
+                }
+            }
+        };
+        let block_sender = PollSender::new(block_sender);
+        let block_stream = Self { mode, provider, block_sender, current_block };
+        Ok((block_stream, block_receiver))
+    }
+}
+
+impl<T> std::fmt::Debug for BlockFetcher<'_, T>
+where
+    RootProvider<T>: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // write every value directly except the future
+        f.debug_struct("BlockFetcher")
+            .field("mode", &self.mode)
+            .field("provider", &self.provider)
+            .field("block_sender", &self.block_sender)
+            .finish()
+    }
+}
+
+/// All error variants for the block stream
+#[derive(Debug, thiserror::Error)]
+pub enum BlockFetcherError {
+    /// The range is empty
+    #[error("The range is empty")]
+    RangeEmpty,
+}
+
+impl<'a, T> Future for BlockFetcher<'a, T>
```

Comment:
@mattsse @Rjected this feels unnecessarily complex, couldn't it just be:

spawn an async loop task that does:

```
let block = provider.get_block_by_number(Latest).await;
if block.is_err() || block == Ok(None) {
   info!("retrying block {self.highest_block}");
   continue; // should this sleep?
}
if block.number > self.highest_block {
  self.highest_block = block;
  block_sender.send_item(Some(block))
}

if Some(self.highest_block) == self.mode.range_end() {
  block_sender.send_item(None) // signal that we're done
}
```

Reply:
Sure, we can rewrite it as an async fn, we can't use `Latest` but that's more of an implementation detail